File: src/main/java/de/esymetric/jerusalem/osmDataRepresentation/osm2ownMaps/longOsm2OwnMaps/LongOsmNodeID2OwnIDMapFileCache.java
package de.esymetric.jerusalem.osmDataRepresentation.osm2ownMaps.longOsm2OwnMaps;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

public class LongOsmNodeID2OwnIDMapFileCache {
	
	int maxSize = 1;  // server: 16  pc: 2
	
	Queue<Integer> queue = new LinkedList<Integer>();
	Map<Integer, LongOsmNodeID2OwnIDMapFile> cache = new HashMap<Integer, LongOsmNodeID2OwnIDMapFile>();
	
	int numberOfAccesses;
	int numberOfFileChanges;
	
	String dataDirectoryPath;
	
	public LongOsmNodeID2OwnIDMapFileCache(String dataDirectoryPath, int maxCacheSize) {
		this.dataDirectoryPath = dataDirectoryPath;
		this.maxSize = maxCacheSize;

		String dirPath = dataDirectoryPath + File.separatorChar + LongOsmNodeID2OwnIDMapFile.DIR;
		//deleteRecursively(dirPath);
		new File(dirPath).mkdir();
	}
	
	public boolean put(long osmNodeID, int ownNodeID) {
		LongOsmNodeID2OwnIDMapFile f1 = getMap(osmNodeID, false);
		return f1.put(osmNodeID, ownNodeID);
		
	}
	

	public int get(long osmNodeID) {
		LongOsmNodeID2OwnIDMapFile f1 = getMap(osmNodeID, true);
		return f1.get(osmNodeID);
	}

	public int getEstimatedMemoryUsedMB() {
		return (int)((long)cache.size() * (long)LongOsmNodeID2OwnIDMapFile.FILE_SIZE / 1024L / 1024L);
	}
	
	
	private LongOsmNodeID2OwnIDMapFile getMap(long osmNodeID, boolean readOnly) {
		numberOfAccesses++;
		
		int fileNumber = LongOsmNodeID2OwnIDMapFile.getFileNumber(osmNodeID);
		
		// get it from cache?
		
		if( cache.containsKey(fileNumber) ) {
			queue.remove(fileNumber);
			queue.add(fileNumber); // insert at end of queue
			return cache.get(fileNumber);
		}
		
		// not in queue
		
		//TODO if( readOnly && !new File(filePath).exists() ) return null;
		
		// remove first element?
		
		LongOsmNodeID2OwnIDMapFile raf = null;
		if( queue.size() >= maxSize ) {
			int fn = queue.poll();
			raf = cache.remove(fn);
			/*
			try {
				raf.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
			*/
		}
		
			//raf = new RandomAccessFile(filePath, readOnly ? "r" : "rw");
			if( raf == null )
			raf = new LongOsmNodeID2OwnIDMapFile(dataDirectoryPath);
			raf.openFile(fileNumber);
			numberOfFileChanges++;
		
		cache.put(fileNumber, raf);
		queue.add(fileNumber);
		
		return raf;
	}
	
	public int getAndClearNumberOfFileChanges() {
		int n = numberOfFileChanges;
		numberOfFileChanges = 0;
		return n;
	}
	
	public void close() {
		for( LongOsmNodeID2OwnIDMapFile raf : cache.values())
			try {
				raf.close();
			} catch (Exception e) {
				e.printStackTrace();
			}
		queue.clear();
		cache.clear();
	}
}

File: src/main/java/de/esymetric/jerusalem/osmDataRepresentation/osm2ownMaps/longOsm2OwnMaps/LongOsmNodeID2OwnIDMapFileCache.kt
package de.esymetric.jerusalem.osmDataRepresentation.osm2ownMaps.longOsm2OwnMaps

import java.io.File
import java.util.*

class LongOsmNodeID2OwnIDMapFileCache(var dataDirectoryPath: String, maxCacheSize: Int) {
    private var maxSize = 1 // server: 16  pc: 2
    private var queue: Queue<Int> = LinkedList()
    private var cache: MutableMap<Int, LongOsmNodeID2OwnIDMapFile> = HashMap()
    private var numberOfAccesses = 0
    private var numberOfFileChanges = 0

    init {
        maxSize = maxCacheSize
        val dirPath = dataDirectoryPath + File.separatorChar + LongOsmNodeID2OwnIDMapFile.Companion.DIR
        File(dirPath).mkdir()
    }

    fun put(osmNodeID: Long, ownNodeID: Int): Boolean {
        val f1 = getMap(osmNodeID, false)
        return f1!!.put(osmNodeID, ownNodeID)
    }

    operator fun get(osmNodeID: Long): Int {
        val f1 = getMap(osmNodeID, true)
        return f1!![osmNodeID]
    }

    val estimatedMemoryUsedMB: Int
        get() = (cache.size.toLong() * LongOsmNodeID2OwnIDMapFile.Companion.FILE_SIZE.toLong() / 1024L / 1024L).toInt()

    private fun getMap(osmNodeID: Long, readOnly: Boolean): LongOsmNodeID2OwnIDMapFile? {
        numberOfAccesses++
        val fileNumber: Int = LongOsmNodeID2OwnIDMapFile.Companion.getFileNumber(osmNodeID)

        // get it from cache?
        if (cache.containsKey(fileNumber)) {
            queue.remove(fileNumber)
            queue.add(fileNumber) // insert at end of queue
            return cache[fileNumber]
        }

        // not in queue

        // remove first element?
        var raf: LongOsmNodeID2OwnIDMapFile? = null
        if (queue.size >= maxSize) {
            val fn = queue.poll()
            raf = cache.remove(fn)
        }
        if (raf == null) raf = LongOsmNodeID2OwnIDMapFile(dataDirectoryPath)
        raf.openFile(fileNumber)
        numberOfFileChanges++
        cache[fileNumber] = raf
        queue.add(fileNumber)
        return raf
    }

    val andClearNumberOfFileChanges: Int
        get() {
            val n = numberOfFileChanges
            numberOfFileChanges = 0
            return n
        }

    fun close() {
        for (raf in cache.values) try {
            raf.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        queue.clear()
        cache.clear()
    }
}